ydb/core/blobstorage/vdisk/balance/deleter.cpp

```diff
@@ -0,0 +1,194 @@
+                    vDiskId, TInstant::Max(), NKikimrBlobStorage::EGetHandleClass::AsyncRead,
+                    TEvBlobStorage::TEvVGet::EFlags::None, i,
+                    {{item.Key.FullID(), 0, 0}}
+                );
+                TReplQuoter::QuoteMessage(
+                    Quoter,
+                    std::make_unique<IEventHandle>(QueueActorMapPtr->at(TVDiskIdShort(vDiskId)), ctx.SelfID, ev.release()),
+                    0
+                );
+                ++ExpectedResponses;
+            }
+            return WAITING_RESPONSES;
+        }
+
+        std::optional<TVector<TPartOnMain>> TryGetResults() {
+            if (ExpectedResponses != 0 && ExpectedResponses == Responses) {
+                ExpectedResponses = 0;
+                Responses = 0;
+                return Result;
+            }
+            return std::nullopt;
+        }
+
+        void Handle(TEvBlobStorage::TEvVGetResult::TPtr ev) {
+            ++Responses;
+            auto msg = ev->Get()->Record;
+            if (msg.GetStatus() != NKikimrProto::EReplyStatus::OK) {
+                return;
+            }
+            ui64 i = msg.GetCookie();
+            auto res = msg.GetResult().at(0);
+            for (ui32 partIdx: res.GetParts()) {
+                if (partIdx == Result[i].Key.PartId()) {
+                    Result[i].HasOnMain = true;
+                }
+            }
+        }
+    };
+
+
+    class TDeleter : public TActorBootstrapped<TDeleter> {
+        TActorId NotifyId;
+        std::shared_ptr<TBalancingCtx> Ctx;
+        TPartsRequester PartsRequester;
+        ui32 OrderId = 0;
+
+        struct TStats {
+            ui32 PartsRequested = 0;
+            ui32 PartsDecidedToDelete = 0;
+            ui32 PartsMarkedDeleted = 0;
+        };
+        TStats Stats;
+
+        void DoJobQuant(const TActorContext &ctx) {
+            auto status = PartsRequester.DoJobQuant(ctx);
+            if (auto batch = PartsRequester.TryGetResults()) {
+                Stats.PartsRequested += batch->size();
+                for (auto& part: *batch) {
+                    if (part.HasOnMain) {
+                        ++Stats.PartsDecidedToDelete;
+                        DeleteLocal(ctx, part.Key, part.Ingress);
+                    }
+                }
+            }
+            LOG_DEBUG_S(ctx, NKikimrServices::BS_VDISK_BALANCING,
+                        Ctx->VCtx->VDiskLogPrefix << "PartsRequester status " << ": " << (ui32)status << " " << Stats.PartsDecidedToDelete);
+            if (status == TPartsRequester::EState::FINISHED && Stats.PartsDecidedToDelete == Stats.PartsMarkedDeleted) {
+                Send(SelfId(), new NActors::TEvents::TEvPoison);
+            }
+        }
+
+        void DeleteLocal(const TActorContext &ctx, const TLogoBlobID& key, const TIngress& ingress) {
+            TLogoBlobID keyWithoutPartId(key, 0);
+
+            TIngress patchedIngress = ingress.CopyWithoutLocal(Ctx->GInfo->GetTopology().GType);
```

Comment:
I don't know where you get it, but the main idea is that every time you put something into database, you construct independent ingress containing the very local patch. In this case this is handoff deletion marker. And nothing more. It'll all be compacted to the single record.